src/cargo/util/lints.rs

```diff
@@ -217,3 +235,113 @@ pub fn check_implicit_features(
     }
     Ok(())
 }
+
+const UNUSED_OPTIONAL_DEPENDENCY: Lint = Lint {
+    name: "unused_optional_dependency",
+    desc: "unused optional dependency",
+    groups: &[],
+    default_level: LintLevel::Warn,
+    edition_lint_opts: None,
+};
+
+pub fn unused_dependencies(
+    pkg: &Package,
+    path: &Path,
+    lints: &TomlToolLints,
+    error_count: &mut usize,
+    gctx: &GlobalContext,
+) -> CargoResult<()> {
+    let edition = pkg.manifest().edition();
+    // This lint is only relevant for editions 2024 and later.
+    if edition <= Edition::Edition2021 {
```

Comment:
Since the comment is referencing Edition 2024, would this be clearer as:

```suggestion
    if edition < Edition::Edition2024 {
```